services/datcore-adapter/src/simcore_service_datcore_adapter/modules/pennsieve.py

```diff
@@ -0,0 +1,378 @@
+                    [
+                        ancestor_pck["content"]["name"]
+                        for ancestor_pck in collection_pck["ancestors"]
+                    ]
+                )
+            )
+            / collection_pck["content"]["name"]
+        )
+
+        return (
+            [
+                FileMetaData.from_pennsieve_package(
+                    pck,
+                    await self._get_package_files(
+                        api_key, api_secret, pck["content"]["nodeId"]
+                    )
+                    if pck["content"]["packageType"] != "Collection"
+                    else [],
+                    base_path=base_path,
+                )
+                for pck in islice(collection_pck["children"], offset, offset + limit)
+            ],
+            len(collection_pck["children"]),
+        )
+
+    async def list_all_dataset_files(
+        self, api_key: str, api_secret: str, dataset_id: str
+    ) -> List:
+        """returns ALL the files belonging to the dataset, can be slow if there are a lot of files"""
+
+        async def _parse_dataset_items() -> List[FileMetaData]:
+            file_meta_data = []
+            cursor = ""
+            page_size = 1000
+
+            # get number of packages
+            num_packages = await self._get_dataset_packages_count(
+                api_key, api_secret, dataset_id
+            )
+
+            # get dataset details
+            dataset_details = await self._get_dataset(api_key, api_secret, dataset_id)
+            base_path = Path(dataset_details["content"]["name"])
+
+            # get all data packages inside the dataset
+            all_packages: Dict[str, Dict[str, Any]] = {}
+            while resp := await self._get_dataset_packages(
+                api_key, api_secret, dataset_id, page_size, cursor
+            ):
+                cursor = resp.get("cursor")
+                all_packages.update(
+                    {p["content"]["id"]: p for p in resp.get("packages", [])}
+                )
+                logger.warning(
+                    "received packages [%s/%s]", len(all_packages), num_packages
+                )
+                if cursor is None:
+                    # the whole collection is there now
+                    break
+
+            # get the information about the files
+            for package_id, package in all_packages.items():
+                if package["content"]["packageType"] == "Collection":
+                    continue
+
+                file_path = base_path / _compute_file_path(all_packages, package)
+                files = await self._get_package_files(
+                    api_key, api_secret, package["content"]["nodeId"]
+                )
+                if len(files) > 1:
+                    logger.warning(
+                        "Package %s contains more than 1 file, only the first one will be retrieved, please check",
+                        package_id,
+                    )
```

Comment:
is there a reason for this?

Reply:
well I don't know. what I have seen is:
- the code was made so
- effectively I have not seen yet a single dataset in datcore where there were more than 1 file in a data package. but since it's possible I guess this might happen.
- Anyway I remove that part to a gather.